### ISSUE TITLE
performance issues

### ISSUE THREAD

**Issue by doughsay** [solnic/virtus#287]
First let me say that this is an amazing gem, I really enjoy working with it and it really fits my needs exactly. But I've just realized that I might not be able to use it due to performance issues, which is a real shame.

Take this rather contrived example below to see how badly it seems to perform for me:

``` ruby
class Foo
  attr_accessor :foo
end

class Bar
  include Mongoid::Document
  field :bar, :type => String
end

class Baz
  include Virtus.model
  attribute :baz, String
end

time_me = ->(proc) {
  start_time = Time.now.to_f
  proc.call()
  end_time = Time.now.to_f
  time_in_millis = ((end_time - start_time) * 1000).to_i
  puts time_in_millis
}

time_me.call ->{1000.times{ foo = Foo.new; foo.foo = 'foo'; }}
# => 0

time_me.call ->{1000.times{ bar = Bar.new; bar.bar = 'bar'; }}
# => 21

time_me.call ->{1000.times{ baz = Baz.new; baz.baz = 'baz'; }}
# => 592
```

As you can see, it appears Virtus models perform 30x slower (than mongoid documents) in simple cases like this, where mongoid is providing the same functionality, i.e. attribute definition and type coercion.

Basically in my case, what I'm seeing is Virtus taking almost 2 seconds to "hydrate" less than 10 models each with about 100 attributes. This makes it pretty much unusable...  I really hope there's something that can be done about it because I'd love to be able to use Virtus.
